### PR TITLE
Implement caching layer in v1

### LINKED ISSUE #342: Implement caching layer
Design doc: https://docs.google.com/document/d/1P6boQoVya126qm0JI7oqyJ24bLdUw1cD7-BrGlxPuUs/edit

Changes:
- [ ] Generate and store `RequestHash` field in Dynamo `RequestTable`
- [ ] Implement cache hit path in `query_runner.py`
- [ ] Update output matrix S3 prefix to include `data_version` and `request_hash`
- [ ] Update GET endpoint to retrieve cached entries

DoD:
- [ ] Requesting the same matrix multiple times (defined by resulting from the same cellkey set) does not generate a new matrix and returns a copy of an existing matrix
- [ ] No regression in v0

### PR DESCRIPTION
**Pull request by calvinnhieu**
Closes #342

### CODE REVIEW

File: matrix/common/query/cell_query_results_reader.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+        Returns:
+            dataframe of cell metadata. Index is "cellkey" and other columns are metadata
+            fields.
+        """
+
+        cell_table_columns = self._map_columns(self.manifest["columns"])
+        cell_table_dtype = {c: "category" for c in cell_table_columns}
+        cell_table_dtype["genes_detected"] = "uint32"
+        cell_table_dtype["cellkey"] = "object"
```
[the remark this answers is not in the document]
Comment (calvinnhieu) [reply matrix/common/query/cell_query_results_reader.py]: Yup, these will be included after I rebase from master

**Comment by calvinnhieu** [comment]
Ah good catch - I missed updating the target path in the matrix converter

File: matrix/docker/query_runner.py [hunk 2]
```diff
@@ -52,6 +60,13 @@ def run(self, max_loops=None):
                     self.redshift_handler.transaction([query], read_only=True)
                     logger.info(f"Finished running query from {obj_key}")
 
+                    if query_type == QueryType.CELL:
+                        cached_result_s3_key = request_tracker.retrieve_cached_result_s3_key()
```
[the remark this answers is not in the document]
Comment (calvinnhieu) [reply matrix/docker/query_runner.py]: Will update permissions in terraform

File: matrix/common/request/request_tracker.py [hunk 3]
```diff
@@ -144,6 +198,26 @@ def initialize_request(self, fmt: str) -> None:
...
+        cell_manifest_key = f"s3://{os.environ['MATRIX_QUERY_RESULTS_BUCKET']}/{self.request_id}/cell_metadata_manifest"
+        reader = CellQueryResultsReader(cell_manifest_key)
+        cell_df = reader.load_results()
+        cellkeys = cell_df.index
+
+        h = hashlib.md5()
+        # TODO: include feature
+        h.update(self.format)
+        for key in cellkeys:
+            h.update(key)
```
[the remark this answers is not in the document]
Comment (calvinnhieu) [reply matrix/common/request/request_tracker.py]: Fixed

File: matrix/docker/query_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (calvinnhieu) [reply matrix/docker/query_runner.py]: Fixed

File: matrix/lambdas/api/v0/core.py [hunk 4]
```diff
@@ -126,17 +126,19 @@ def get_matrix(request_id: str):
...
+                              f"{request_tracker.s3_results_prefix}/{request_id}.{format}.zip"
 
         return ({'request_id': request_id,
                  'status': MatrixRequestStatus.COMPLETE.value,
                  'matrix_location': matrix_location,
                  'eta': "",
                  'message': f"Request {request_id} has successfully completed. "
                             f"The resultant expression matrix is available for download at "
-                            f"{matrix_location}"},
+                            f"{matrix_location} for 30 days."},
```
Comment (mckinsel) [inline matrix/lambdas/api/v0/core.py]: What happens if someone makes a request, waits 29 days, and then does a GET? I think they're get this 30 day message even though the object is about to be deleted.

Comment (calvinnhieu) [reply matrix/lambdas/api/v0/core.py]: Good call. The expiration date (set on object creation as the creation date + bucket expiration policy lifetime) can be retrieved via [head_object](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_object) and can provide a correct expiration.

However, am removing this change for now and will reintroduce when expiration is implemented.

**Comment by mckinsel** [comment]
Hey so what happens if someone makes a request for an mtx matrix and then later makes an identical request except for a loom matrix? Will they just get the mtx again on the second request?

**Comment by calvinnhieu** [comment]
Nope, it'll miss the cache and go through the full request flow. `generate_request_hash` encodes `feature`, `format` and all `cellkeys`

**Comment by mckinsel** [comment]
Ah okay, I overlooked the `format` in `generate_hash`, thanks

**Pull request merged**